proof/refine/ARM/Reply_R.thy

```diff
@@ -96,4 +95,487 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
+  "replyRemoveTCB tptr \<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  unfolding replyRemoveTCB_def decompose_list_refs_of_replies'
+  supply if_cong[cong]
+  apply (wpsimp wp: cleanReply_list_refs_of_replies' hoare_vcg_if_lift hoare_vcg_imp_lift' gts_wp'
+                    haskell_assert_wp
+              simp: pred_tcb_at'_def
+         split_del: if_split)
+  unfolding decompose_list_refs_of_replies'[symmetric] protected_sym_refs_def[symmetric]
+  \<comment>\<open> opt_mapE will sometimes destroy the @{term "(|>)"} inside @{term replyNexts_of}
+      and @{term replyPrevs_of}, but we're using those as our local normal form. \<close>
+  supply opt_mapE[rule del]
+  apply (intro conjI impI allI)
+       \<comment>\<open> Our 6 cases correspond to various cases of @{term replyNext} and @{term replyPrev}.
+           We use @{thm ks_reply_at'_repliesD} to turn those cases into facts about
+           @{term replyNexts_of} and @{term replyPrevs_of}. \<close>
+       apply (all \<open>clarsimp simp: isHead_to_head
+                   , normalise_obj_at'
+                   , (drule obj_at'_to_ks_obj_at', clarsimp)+\<close>)
+       apply (all \<open>drule(1) ks_reply_at'_repliesD[folded protected_sym_refs_def]
+                   , clarsimp simp: projectKO_reply isHead_to_head\<close>)
+       \<comment>\<open> Now, for each case we can blow open @{term sym_refs}, which will give us enough new
+           @{term "(replyNexts_of, replyPrevs_of)"} facts that we can throw it all at metis. \<close>
+       apply (clarsimp simp: sym_refs_def split_paired_Ball in_get_refs
+              , intro conjI impI allI
+              ; metis sym_refs_replyNext_replyPrev_proj[folded protected_sym_refs_def] option.sel
+                      option.inject)+
+  done
+
+(* FIXME RT: move to...? *)
+lemma objBits_sc_only_depends_on_scRefills:
+  fixes sc :: sched_context
+    and upd :: "sched_context \<Rightarrow> sched_context"
+  assumes [simp]: "scRefills (upd sc) = scRefills sc"
+  shows "objBits (upd sc) = objBits sc"
+  apply (clarsimp simp: objBits_def objBitsKO_def)
+  done
+
+lemma replyRemoveTCB_valid_idle':
+  "\<lbrace>valid_idle' and valid_pspace'\<rbrace>
+   replyRemoveTCB tptr
+   \<lbrace>\<lambda>_. valid_idle'\<rbrace>"
+  unfolding replyRemoveTCB_def
+  apply (wpsimp wp: cleanReply_valid_pspace' updateReply_valid_pspace' hoare_vcg_if_lift
+                    hoare_vcg_imp_lift' gts_wp' setObject_sc_idle' haskell_assert_wp
+         split_del: if_split)
+  apply (clarsimp simp: valid_reply'_def pred_tcb_at'_def)
+  apply (intro conjI impI allI
+         ; simp?
+         ; normalise_obj_at'?)
+  \<comment>\<open> Weird indentation: around 20 subgoals, almost all of which are trivial/boring facts or
+      contradictions about being a valid SC. \<close>
+  apply (all \<open>(solves \<open>clarsimp simp: valid_idle'_def idle_tcb'_def obj_at'_def isReply_def\<close>)?\<close>)
```

Comment:
I know it'd make the previous block a bit more unwieldy, but did you wanna put this `all` bit into it too? Does it actually solve the 20 subgoal problem?

Reply:
It solves all but 4, hence the indentation afterwards. Not sure which is preferable?

Reply:
If it's a choice between another ; and the `all < blah >`, then I'd go with another ;